[OKTA-459415] First pass at Custom SSPR Guide for dotnet (#3069)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/oie-embedded-sdk-use-case-custom-pwd-recovery-mfa/main/aspnet/siw.md

### 1: Initiate password recovery

To begin the password recovery flow, the user must

> 1. Click the **Forgot Password?** link on the login page.

Reviewer: ```suggestion
1. Click the **Forgot Password?** link on the sign-in page.
```

Author: fixed

> 2. Enter their **Email or Username** in the textbox and click **Next**.

Reviewer: ```suggestion
2. Enter their **Email or Username** in the box and click **Next**.
```

> 3. Choose **Email** as the authenticator to be used for password recovery and click **Submit**

Reviewer: ```suggestion
3. Choose **Email** as the authenticator they want to use for password recovery and click **Submit**.
```

Author: fixed

Okta then tells the user to either click the link in the email or enter the code to continue and sends an email to their email address matching the Forgot Password template that was altered earlier.

<div class="common-image-format">

> ![Screenshot of email sent to user](/img/advanced-use-cases/custom-pwd-recovery-custom-email.png)

Reviewer: see my comment above about updating the image description.

Author: fixed

</div>

The **Reset Password** link in the email includes the `OTP` and `request.relayState` variables as query parameters back to the application. For example,

`https://localhost:44314/magiclink/callback?otp=${oneTimePassword}&state=${request.relayState}` becomes `https://localhost:44314/magiclink/callback?otp=726009&state=1b34371af02dd31d2bc4c48a3607cd32`.

> ### 2: Handle the otp and state parameters

Reviewer: ```suggestion
### 2: Handle the OTP and state parameters
```

Author: Same note as above. The otp parameter is lower case, so this is too.

> Create a callback handler method that takes the `otp` and `state` parameters in the query string and passes them as session parameters to a page containing the Sign-In Widget.

Reviewer: ```suggestion
Create a callback handler method that takes the `otp` and `state` parameters in the query string and passes them as session parameters to a page that contains the Sign-In Widget.
```

```csharp
public async Task<ActionResult> Callback(string state, string otp, string error = null, string error_description = null)
{
    if (!string.IsNullOrEmpty(error) || !string.IsNullOrEmpty(error_description))
    {
        return View(new MagicLinkCallbackModel { Message = $"{error}: {error_description}" });
    }

    SignInWidgetConfiguration siwConfig = Session["siwConfig"] as SignInWidgetConfiguration;

    if (siwConfig != null)
    {
        Session["otp"] = otp;
        Session["state"] = state;
        return View("SignInWidget", siwConfig);
    }
```

> If the otp and state values are not valid or the user is in a different browser after clicking the link, you should advise the user to return to the original tab in the browser where they requested a password reset and enter the otp value there to proceed.

Reviewer: ```suggestion
If the OTP and state values aren't valid or the user is in a different browser after clicking the link, you should advise the user to return to the original tab in the browser where they requested a password reset and enter the OTP value to proceed.
```

Author: fixed

```csharp
    return View(new MagicLinkCallbackModel { Message = $"Please enter the OTP '{otp}' in the original browser tab to finish the flow." });
}
```

> ### 3: Setup and render Widget with otp and State

Reviewer: ```suggestion
### 3: Set up and render widget with OTP and state
```

Author: fixed

> Consume the session `state` and `otp` variables in a Razor page containing the Sign-In Widget.

Reviewer: ```suggestion
Consume the session `state` and `otp` variables in a Razor page that contains the Sign-In Widget.
```

Author: fixed

```razor
@model SignInWidgetConfiguration

@section head
{
    <script src="https://global.oktacdn.com/okta-signin-widget/6.0.0/js/okta-sign-in.min.js" type="text/javascript"></script>
    <link href="https://global.oktacdn.com/okta-signin-widget/6.0.0/css/okta-sign-in.min.css" type="text/css" rel="stylesheet" />
}

<div id="okta-signin-widget-container"></div>
<script type="text/javascript">
    var widgetConfig = @Html.Raw(JsonConvert.SerializeObject(Model));

    @Html.Raw($"widgetConfig.otp = '{Session["otp"]}';widgetConfig.state = '{Session["state"]}';");

    console.log(widgetConfig.interactionHandle);

    const signIn = new OktaSignIn({
        el: '#okta-signin-widget-container',
        ...widgetConfig
    });
    signIn.showSignInAndRedirect()
        .catch(err => {
            console.log('Error happen in showSignInAndRedirect: ', err);
        });
</script>
```

### 4: Display password reset page and continue the password recovery flow

> Once the widget is loaded, it checks whether the state and OTP are valid with Okta. Assuming they are, either the following reset page is displayed, or a prompt is made for the OTP code to be entered. Once the OTP code is entered, the reset page is then displayed. The user continues the password recovery flow described in the [User password recovery guide](/docs/guides/oie-embedded-sdk-use-case-pwd-recovery-mfa/aspnet/main/).

Reviewer: ```suggestion
After the widget is loaded, it checks whether the state and OTP are valid with Okta. Assuming they are, either the following reset page appears, or a prompt appears for the user to enter the OTP code. After the user enters the OTP code, the reset page appears. The user continues the password recovery flow described in the [User password recovery guide](/docs/guides/oie-embedded-sdk-use-case-pwd-recovery-mfa/aspnet/main/).
```

Author: Fixed

<div class="common-image-format">

> ![Screenshot of password reset page](/img/advanced-use-cases/dotnet-custom-pwd-recovery-custom-siw-reset-pwd-page.png)

Reviewer: see my comment above about updating the image description.
